pyphi/tpm.py

```diff
@@ -50,11 +50,35 @@ def condition_tpm_nb(
     tpm, fixed_nodes, state, num_states_per_node=None, node_labels=None
 ):
     df = tpm2df(tpm, num_states_per_node, node_labels)
-    for c in fixed_nodes:
-        df = df.iloc[df.index.get_level_values(c) == state[node_labels.index(c)]]
-    free_nodes = list(set(node_labels) - set(fixed_nodes))
+    num_states_per_node = list(num_states_per_node)
+
+    for c in reversed(fixed_nodes):
+        bool_array = [value == state[c] for value in df.index.get_level_values(c)]
+        df = df.iloc[bool_array]
+        del num_states_per_node[c]
+    
+    fixed_node_labels = [node_labels[fixed_node] for fixed_node in fixed_nodes]
+
+    # TODO This, and probably more of the method, might be improvable
+    # if working with just the NodeLabel class instead of lists here
+    # However, subtracting sets was giving poor ordering, causing errors
+    # downstream, so for now subtracting lists works.
+    def subtract_nodes(a, b):
+        a, b = list(a), list(b)
+        for i in b:
+            if i in a:
+                a.remove(i)
+        return a
+
+    free_nodes = subtract_nodes(node_labels, fixed_node_labels)
```

Comment:
Could you instead work with node indices first, and then convert to labels? This would allow just doing set subtraction and then sorting:
```python
sorted(set(all_nodes) - set(fixed_nodes))
```